### ISSUE TITLE
[Improvement] Disallow sendShuffleData if requireBufferId expired

### ISSUE THREAD

**Issue by xianjingfeng** [apache/incubator-uniffle#76]
We found shuffle server which under high load is easy encounter `java.lang.OutOfMemoryError: Java heap space` even we allocate more jvm heap memory and less `rss.server.buffer.capacity `

The steps for the exception above:
1. When  shuffle server under high load, `requireBufferId` is easy to expire, and Suffle server release `usedMemory`
2. Client `sendShuffleData` using a expired  `requireBufferId`, 
3. Suffle server recive shuffle data and store _in rpc queue_(this part of memory usage was not be added to `usedMemory`)
4. Other clients `requireBuffer`  because `usedMemory` is enough

**Comment by jerqi** [comment]
Do you have the solution of the problem?

**Comment by xianjingfeng** [comment]
Yes, it is be testing in our production environment. I will watch it for a while. If it's OK, I will create a pr

**Comment by jerqi** [comment]
Could you share your solution?  We can discuss first.

**Comment by jerqi** [comment]
cc @colinmjj . There seems not be cases in our production environment. But I think the analysis is correct. What do you think?

**Comment by colinmjj** [comment]
I think @xianjingfeng is right, with current implementation, OOM will happen if `requireBufferId` was expired in Shuffle Server already, this maybe caused by GC, network problem, high workload in shuffle server etc.
It's better to have the limitation to accept the data with `requireBufferId` only to avoid such problem.

**Comment by jerqi** [comment]
closed by #157
